### DISCUSSION TITLE
real time plotting

### DISCUSSION THREAD

**Discussion by Xaego** [h2oai/wave#2117] [Q&A]
Could you help me make real time plotting with H2O wave?
I can do it static, but how to make it realtime?

```py
import time

import numpy as np
import pandas as pd
from h2o_wave import site, ui, data, app, Q


def gen_random_array():
    num_rows = 3000
    df = pd.DataFrame(dict(
        pix=range(num_rows),
        intensity=np.random.random(num_rows)
    ))
    return df

df = gen_random_array()

page = site["/random_monitor"]

plot_1 = page.add("plot_intensity(pix)", ui.plot_card(
    box="1 1 6 4",
    title="Line",
    data=data(
        # fields=df.columns.tolist(),
        # rows=df.values.tolist(),
        fields="={{df.columns.tolist()}}",
        rows=dict(pix=df["pix"], intensity=df["intensity"]),
    ),
    plot=ui.plot(marks=[ui.mark(
        type='line',
        x='=pix', x_title='Length',
        y='=intensity', y_title='Width', shape='circle',
    )]
    )
))

while True:
    df = gen_random_array()
    plot_1.data.usage = df

    pix = df["pix"]
    intensity = df["intensity"]

    plot_1.data.pix = pix
    plot_1.data.intensity = intensity
    plot_1.plot_data[-1] = [pix, intensity]

    page.save()
    time.sleep(1)
```

**Comment by mturoci** [comment]
Hi @Xaego.

Unfortunately, your code is throwing an error for me:

```
Traceback (most recent call last):
  File "/Users/martinturoci/go/src/github.com/h2oai/telesync/py/examples/hello_world.py", line 37, in <module>
    page.save()
  File "/Users/martinturoci/go/src/github.com/h2oai/telesync/py/h2o_wave/h2o_wave/core.py", line 607, in save
    p = self._diff()
        ^^^^^^^^^^^^
  File "/Users/martinturoci/go/src/github.com/h2oai/telesync/py/h2o_wave/h2o_wave/core.py", line 552, in _diff
    d = marshal(dict(d=self._changes))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martinturoci/go/src/github.com/h2oai/telesync/py/h2o_wave/h2o_wave/core.py", line 1138, in marshal
    return json.dumps(d, allow_nan=False, separators=(',', ':'))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martinturoci/.asdf/installs/python/3.11.3/lib/python3.11/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/Users/martinturoci/.asdf/installs/python/3.11.3/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martinturoci/.asdf/installs/python/3.11.3/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/Users/martinturoci/.asdf/installs/python/3.11.3/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type Series is not JSON serializable
```

So I would focus on fixing it first.

Apart from that, I suggest having a look at https://wave.h2o.ai/docs/buffers/ to understand how Wave data buffers work. Your code currently uses `fixed` buffer. However, for realtime updates, you most likely want to use `cyclic` buffer (FIFO) instead so that all you need to do is append new data and drop the old one.

You will need `plot_1.plot_data[-1] = [pix, intensity]` once `cyclic_buffer` is configured. However, both `pix` and `intensity` need to be primitive values as the line corresponds to a single tabular row within the dataset.

**Reply by Xaego** [reply]
Hi @mturoci 
thank you. I still fight with it.

could you clarify if its possible to make program which I needed with H2o wave?
because I tried different dashboard lib and after 2 week understand its impossible...

I need program which is can:
0 work on local relative weak pc without network. like Lenovo ThinkCentre M700 Tiny
1 take real time 2d data from device.
2 plot it each 0.3-0.5 sec. subplot_1
3 have subplot_2 for sum all values of plot from 1 vs iteration. also real time
4 have button which stop current measurement and send commands to arduino and my sensor.
5 can still make realtime plots after step 4
6 have button for closing connection to arduino and my sensor

**Reply by mturoci** [reply]
Yes, totally possible. However, if you want to also handle user interactions (button clicks), you will need to use [apps](https://wave.h2o.ai/docs/apps/), not [scripts](https://wave.h2o.ai/docs/scripts).

Based on what you describe it seems like what you want is a background job that reads the data from arduino and plots them and then a simple UI for controlling the arduino around that. Have a look at https://medium.com/@unusualcode/background-jobs-in-wave-or-how-not-to-kill-your-ui-ae1fed95693a to learn more about how to start background jobs, update Wave UI and stop them.
